File: foobar.py
#!/usr/bin/env python3

import sys

G_RECUSRION_LIMIT = 2**11

sys.setrecursionlimit(G_RECUSRION_LIMIT)

MSG_FOO = "Foo"
MSG_BAR = "Bar"

def foo_bar(counter, maximum):
    if counter > maximum:
        return
    if counter % 3 == 0:
        msg = MSG_BAR
    else:
        msg = MSG_FOO
    print('{:d} {:s}'.format(counter, msg))
    foo_bar(counter + 1, maximum)

if __name__ == "__main__":
    maximum = 0
    if len(sys.argv) > 1:
        maximum = int(sys.argv[1])
    else:
        print("Input number greater than 0: ")
        maximum = int(input())
    if maximum < 1:
        print("Incorrect input! Number must be greater than 0.")
    elif maximum >= G_RECUSRION_LIMIT:
        print("Sorry, too big value. Please use numbers up to {:d}. It is recursion limit.".format(G_RECUSRION_LIMIT))
    else:
        foo_bar(1, maximum)

File: foobar.c
#include <stdint.h>
#include <stdio.h>
#include <errno.h>

const char msg_foo[] = "Foo";
const char msg_bar[] = "Bar";

static void foo_bar(uint32_t maximum, uint32_t counter)
{
	if (counter > maximum) return;
	const char *msg = (counter % 3) ? msg_foo : msg_bar;
	printf("%u %s\n", counter, msg);
	++counter;
	foo_bar(maximum, counter);
}

int main(int argc, char* argv[])
{
	int status;
	uint32_t maximum;

	// If run without arguments
	if (argc == 1) {
		puts("Input number greater than 0: ");
		status = scanf("%u", &maximum);
	}
	else {
		status = sscanf(argv[1], "%u", &maximum);
	}
	// WARNING: Unsigned integer overflow not handled
	if (status < 1 || maximum < 1) {
		puts("Incorrect input\n");
		return -1;
	}
	
	foo_bar(maximum, 1);
	
	return 0;
}
